### ISSUE TITLE
RPi.GPIO is not a pure Python 3 wheel

### ISSUE THREAD

**Issue by jaegersb** [pyscript/pyscript#672]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a solution or duplication
- [X] I already searched in Google and didn't find any good information or help

### What happened?

I created a script that reads 3 pins from a Raspberry Pi  and prints them (Low, Med, High), e.g.
if GPIO.input(11) == GPIO.HIGH: print ('Low')

To pull the input I loaded the latest stable RPi.GPIO wheel using py-env
<py-env>
- './RPi.GPIO-0.7.1-cp39-cp39-linux_armv7l.whl'
</py-env>

When calling the html with pyscript it loads and gets stuck at Initializing Components...
<img width="441" alt="Screenshot 2022-08-09 at 09 23 51" src="https://user-images.githubusercontent.com/107865614/183589419-7cc68f2c-4c93-431a-ae22-6ca0147707d1.png">

Loading the local module RPi.GPIO and a data.py file as in the tutorial does not work either, I tried numpy with the same syntax and it executed. So I assume it's an issue with the RPi.GPIO which works fine on Python 3.9 locally.

### Console info

```shell
stores.ts:32 adding initializer async ƒ mountElements() {
        console.log('Collecting nodes to be mounted into python namespace...');
        const pyodide = await pyodideReadyPromise$1;
        const matches = document.querySelectorAl…
stores.ts:34 adding initializer async ƒ mountElements() {
        console.log('Collecting nodes to be mounted into python namespace...');
        const pyodide = await pyodideReadyPromise$1;
        const matches = document.querySelectorAl…
stores.ts:40 adding post initializer async ƒ initHandlers() {
        console.log('Collecting nodes...');
        const pyodide = await pyodideReadyPromise$1;
        for (const pysAttribute of pysAttributeToEvent.keys()) {
            await cr…
stores.ts:42 adding post initializer async ƒ initHandlers() {
        console.log('Collecting nodes...');
        const pyodide = await pyodideReadyPromise$1;
        for (const pysAttribute of pysAttributeToEvent.keys()) {
            await cr…
pyenv.ts:10 RUNTIME READY
pyconfig.ts:17 config set!
pyconfig.ts:22 initializers set
pyconfig.ts:27 post initializers set
pyconfig.ts:32 post initializers set
pyconfig.ts:37 post initializers set
pyconfig.ts:32 post initializers set
pyscript.ts:66 connected
stores.ts:32 adding initializer async ƒ loadEnv() {
                await loadPackage(env);
                console.log('environment loaded');
            }
pyconfig.ts:22 initializers set
stores.ts:34 adding initializer async ƒ loadEnv() {
                await loadPackage(env);
                console.log('environment loaded');
            }
stores.ts:32 adding initializer async ƒ loadPaths() {
                for (const singleFile of paths) {
                    console.log(`loading ${singleFile}`);
                    try {
                        await loadFromFile(singleFi…
pyconfig.ts:22 initializers set
stores.ts:34 adding initializer async ƒ loadPaths() {
                for (const singleFile of paths) {
                    console.log(`loading ${singleFile}`);
                    try {
                        await loadFromFile(singleFi…
pyenv.ts:60 environment loading... Array(1)
pyconfig.ts:17 config set!
pyconfig.ts:112 config set Object
pyconfig.ts:124 Initializing runtimes...
interpreter.ts:5 creating pyodide runtime
favicon.ico:1          Failed to load resource: the server responded with a status of 404 (Not Found)
pyodide.asm.js:14 Python initialization complete
interpreter.ts:15 loading micropip
load-package.ts:329 Loading micropip, pyparsing, packaging, distutils
load-package.ts:389 Loaded micropip, packaging, distutils, pyparsing
interpreter.ts:17 loading pyscript...
interpreter.ts:30 done setting up environment
pyenv.ts:10 RUNTIME READY
pyscript.ts:148 Collecting nodes to be mounted into python namespace...
pyodide.asm.js:14 Uncaught (in promise) PythonError: Traceback (most recent call last):
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/lib/python3.10/asyncio/tasks.py", line 234, in __step
    result = coro.throw(exc)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 183, in install
    transaction = await self.gather_requirements(requirements, ctx, keep_going)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 173, in gather_requirements
    await gather(*requirement_promises)
  File "/lib/python3.10/asyncio/futures.py", line 284, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
    future.result()
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/lib/python3.10/asyncio/tasks.py", line 232, in __step
    result = coro.send(None)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 243, in add_requirement
    raise ValueError(f"'{wheel['filename']}' is not a pure Python 3 wheel")
ValueError: 'RPi.GPIO-0.7.1-cp39-cp39-linux_armv7l.whl' is not a pure Python 3 wheel

    at new_error (pyodide.asm.js:14:238191)
    at pyodide.asm.wasm:0xedbcb
    at pyodide.asm.wasm:0xf1a0e
    at method_call_trampoline (pyodide.asm.js:14:238105)
    at pyodide.asm.wasm:0x134c2c
    at pyodide.asm.wasm:0x217a84
    at pyodide.asm.wasm:0x174a14
    at pyodide.asm.wasm:0x135149
    at pyodide.asm.wasm:0x135243
    at pyodide.asm.wasm:0x1352e6
    at pyodide.asm.wasm:0x1fff83
    at pyodide.asm.wasm:0x1f98b5
    at pyodide.asm.wasm:0x135329
    at pyodide.asm.wasm:0x201f1b
    at pyodide.asm.wasm:0x1ff9ff
    at pyodide.asm.wasm:0x1f98b5
    at pyodide.asm.wasm:0x135329
    at pyodide.asm.wasm:0xf16d8
    at Object.Module.callPyObjectKwargs (pyproxy.gen.ts:360:23)
    at Object.Module.callPyObject (pyproxy.gen.ts:384:17)
    at wrapper (pyodide.asm.js:14:205222)
```

**Comment by antocuni** [comment]
The error message that you are getting is correct: `RPi.GPIO` is *not* a pure python wheel and thus cannot be installed by pyscript, as explained by the [Pyodide docs](https://pyodide.org/en/stable/usage/loading-packages.html). Look at the name of the wheel:
```
RPi.GPIO-0.7.1-cp39-cp39-linux_armv7l.whl
```
The suffix `-linux_armv7l` means that it's a platform specific wheel compiled for ARM CPUs and for linux. PyScript/Pyodide work in the browser, the they use the virtual "WASM" CPU and no operating system.

But apart from the technical details, I fear that what you are trying to do cannot be done at all. PyScript apps are executed entirely inside the browser, which is a sandboxed environment and which cannot access the underlying hardware: so, there is no chance that you can control the GPIO pins of a raspberry from PyScript.

A good rule of thumbs is: if it cannot be done with javascript, it cannot be done with pyscript either.
(Note: I know that you can control GPIOs with javascript, but only if you run it with node.js, i.e. *outside* the browser: this is the equivalent of running your python script directly from the console).

I fear we will need to close this issue as `wontfix`, but I'll let you answer in case I misunderstood something. Also @ntoll might be interested in the subject.

**Comment by ntoll** [comment]
Hi.

My colleague @antocuni is correct. The author of the GPIO module is a friend of mine, and I can confirm it is not pure Python and targets the RPi platform only (which means it won't work in the browser). Antonio's point about the sandboxed nature of the browser environment also makes control of the GPIO pins impossible.

Sorry to be the bearer of bad news.

**Comment by jaegersb** [comment]
Thanks for the explanation. I am only trying to read GPIO status, not control them.
It works using Flask (even to control), but takes many lines of code. 
I hoped I could cut things short, but then I'll stay on the Flask for now and check in future releases here.

Thanks - JJ

**Comment by antocuni** [comment]
> Thanks for the explanation. I am only trying to read GPIO status, not control them.

that's irrelevant. From the point of view of the browser, the GPIO pins don't even exist, so both reading and writing is impossible.

> It works using Flask (even to control), but takes many lines of code. 

yes, in Flask it works because in that case your python code runs directly on the raspberry pi, using the "stock python" instead of PyScript. Then, the flask code generates HTML which is then sent to the browser for display, but the actualy GPIO-controlling is made *outside* the browser

> I hoped I could cut things short, but then I'll stay on the Flask for now and check in future releases here.

PyScript will be very happy to support GPIO _if they are supported by the browser_. 
Until then, it's plainly impossible.

By googling around, I found a couple of old W3C proposals which discussed the matter. I don't know what is the status, but I guess it's not going to happen any time soon:
https://browserobo.github.io/WebGPIO/
https://www.w3.org/community/browserobo/

Closing at at wontfix
